Дополнено функциональность хранения истории задач

Sayders11 opened the pull request:
1) Теперь история хранит неограниченное количество просмотров
2) Одинаковые просмотры не дублируются
3) Удаление и добавление просмотра в историю происходит за константное время

Makhoro committed:
Добавил новую реализацию хранения истории:
1) История не ограничена размером
2) Удаление происходит за константное время
3) Изменил реализацию метода getHistory() в InMemoryHistoryManager

Makhoro committed:
1) Покрыл HistoryManager тестами
2) Изменил модель методов HistoryManager на объектную, геттеры из TaskManager теперь возвращают объект по id

Makhoro committed:
1) Покрыл HistoryManager тестами
2) Изменил модель методов HistoryManager на объектную, геттеры из TaskManager теперь возвращают объект по id

LexLippi commented on Test/service/HistoryManagerTest.java:
```diff
@@ -0,0 +1,76 @@
+class HistoryManagerTest {
+    InMemoryHistoryManager hManager = new InMemoryHistoryManager();
+    InMemoryTaskManager tManager = new InMemoryTaskManager(hManager);
+    private Task task1;
```
В тестах модификаторы доступа не нужны

Sayders11 replied:
Поправил

LexLippi commented on Test/service/HistoryManagerTest.java:
```diff
@@ -0,0 +1,76 @@
+    private Subtask subtask1;
+    private Subtask subtask2;
+
+    @BeforeEach
+    void init() {
+        Task task1 = new Task("task1", "taskDescription");
+        this.task1 = task1;
+        Task task2 = new Task("task2", "taskDescription");
+        this.task2 = task2;
+        Epic epic1 = new Epic("epic1", "epicDescription");
+        this.epic1 = epic1;
+        Epic epic2 = new Epic("epic2", "epicDescription");
+        this.epic2 = epic2;
+        Subtask subtask1 = new Subtask("subtask1", "subtaskDescription", epic1);
+        this.subtask1 = subtask1;
+        Subtask subtask2 = new Subtask("subtask2", "subtaskDescription", epic2);
+        this.subtask2 = subtask2;
+
+        tManager.createTask(task1);
+        tManager.createTask(task2);
+        tManager.createEpic(epic1);
+        tManager.createEpic(epic2);
+        tManager.createSubtask(subtask1);
+        tManager.createSubtask(subtask2);
+    }
```
Зачем тебе здесь `tManager` и этот метод, если ты его нигде в тестах не используешь?

Sayders11 replied:
Айди у объекта инициализируется только после его создания через таск-менеджер, поэтому для тестов приходится создавать его (tManager) и уже в него добавлять таски, чтобы у них появилась айдишка

LexLippi commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -3,24 +3,86 @@
+        }
+    }
+
+    HashMap<Integer, Node> history = new HashMap<>();
+    Node first;
+    Node last;
```
Для соблюдения инкапсуляции данным полям необходимы модификаторы доступа

Sayders11 replied:
Поправил

LexLippi commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -3,24 +3,86 @@
+    Node last;
+
+    private void linkLast(Task task) {
+        Node l = last;
```
Лучше не использовать однобуквенные названия переменных

Sayders11 replied:
Поправил

LexLippi commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -3,24 +3,86 @@
+
+    private void linkLast(Task task) {
+        Node l = last;
+        Task e = task;
```
Зачем создавать эту переменную? Почему бы сразу не использовать `task`?

Sayders11 replied:
На моменте, когда все писал, старался прописывать код максимально подробно для себя, чтобы лучше всё понять и меньше путаться.
В следующий раз постараюсь все подчищать перед коммитом)

LexLippi commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -3,24 +3,86 @@
+        history.remove(id);
+    }
+
+    public ArrayList<Task> getTasks() {
```
Методы не задекларированные в интерфейсе не должны быть публичными

Sayders11 replied:
Поправил

LexLippi commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -3,24 +3,86 @@
+    }
+
+    private void removeNode(int id) {
+        Node current = history.get(id);
```
Если задачи с таким идентификатором не существует, то в следующей строчке произойдет `NullPointerException` -- этого хотелось бы избежать

Sayders11 replied:
Поправил, наверное) Сейчас как раз будем проходить исключения, попробую потом сделать получше, если этот вариант не очень

LexLippi commented on src/service/InMemoryHistoryManager.java:
```diff
@@ -3,24 +3,86 @@
+            linkLast(task);
+        } else {
+            linkLast(task);
+            history.put(task.getId(), last);
```
Положить задачу в `Map` надо в обоих случаях, так что эту строчку можно вынести из условного оператора, как и `linkLast`

Sayders11 replied:
Поправил

LexLippi reviewed (changes requested):
Привет!

Код я посмотрел, но на будущее, пока все проверки не пройдены, я могу не смотреть код.

Как минимум одна из проблем связана с тем, что папка с тестами называется `Test` с заглавной буквы. Если будут еще проблемы -- обращайся к наставнику.

Помимо этого есть некоторое количество замечаний по коду, которые необходимо исправить!

Желаю удачи!

Makhoro committed:
1) Исправлены недочеты по модификаторам
2) Подчищены лишние строки кода
3) Поправил название папки с тестами
4) Добавил отлов исключения в removeNode

LexLippi commented:
Привет!

Отличная работа!

Желаю удачи в следующих спринтах!
